### ISSUE TITLE
Cut-Introduction misses some grammars

### ISSUE THREAD

**Issue by newca12** [newca12/gapt#262]
_From [stefan.hetzl](https://code.google.com/u/stefan.hetzl/) on March 21, 2014 08:42:18_

Do the following in the CLI:

:load ../examples/ProofSequences.scala
val p = LinearExampleProof( 9 )
val T = extractTerms( p )
val Gs = computeGrammars( T )

Then Gs contains 14 grammars and a call e.g. to

seeNFirstGrammars( Gs, 14 )

shows that the smallest among them is of size 7.

However, there is a smaller grammar. Note that T = { f_1(0),...,f_1(s^8(0)) }. A grammar of size 6 which represents T is:

{ α_0, s^3(α_0), s^6(α_0) } o { 0, s(0), s(s(0)) }

Another grammar of size 6 which represents T is:

{ α_0, s(α_0), s(s(α_0)) } o { 0, s^3(0), s^6(0) }

Both of these grammars are not found but should be.

\--

Similar things seem to happen also with other of the example proof sequences.

This looks related to the build problem after commit r1917 .

_Original issue: http://code.google.com/p/gapt/issues/detail?id=262_

**Comment by newca12** [comment]
_From [janos.ta...@gmail.com](https://code.google.com/u/110287254390562270748/) on March 21, 2014 14:27:15_

I compared the outputs for cutIntro/cutIntroG in 1916 and 1917 and there are some strange, non-deterministic effects:

v1916:
\> val p = LinearExampleProof(9)
\> cutIntro(p)

Number of grammars: 25
Smallest grammar-size: 8
Number of smallest grammars: 24
(...)
at.logic.lk.base.LKUnaryRuleCreationException

\> cutIntro(p)

Number of grammars: 23
Smallest grammar-size: 7
Number of smallest grammars: 1
(...)
\<success>

v1917:

\>val p = LinearExampleProof(9)
\>cutIntro(p, NoConstraint)     //this was cutIntroG in v1916

Number of grammars: 15
Smallest grammar-size: 6
Number of smallest grammars: 1



It seems that this symptom has been occurring since before v1917 and even in the regular cutIntro. In particular, it seems that non-minimal grammars result in an at.logic.lk.base.LKUnaryRuleCreationException, which I've never seen before. Perhaps this is also the reason for the error which Alex mentioned last week w.r.t. cutIntro not finding a minimal grammar but working fine after recompilation.

Since the cutIntro-code has been relative stable for the last few weeks, could it be some other component causing this behavior?

**Comment by newca12** [comment]
_From [janos.ta...@gmail.com](https://code.google.com/u/110287254390562270748/) on March 31, 2014 02:22:22_

After a bit of forensics, I tracked down the source of the error.

The error was exposed (not caused) by the expansion sequents. The term extraction from them returns the terms of the proofs in a random order, whereas the term extraction from the LK proofs was completely deterministic. This in itself isn't erroneous behavior, but it exposed a conceptual flaw in the delta table.

The keys of the delta table are lists, not sets, and this causes functionally identical keys to create different lines simply by virtue of ordering. An example of this are these two lines, generated during a compression of LinearExampleProof(4):

[s(0), 0] -> {s²(alpha_0), {3,2}}
[0, s(0)] -> {alpha_0,    {0,1}}

Because of this spurious splitting, no decompositions were found.

This was probably an issue well before the expansion sequents, but I suspect that the terms of the example sequences were simply returned in a "good" order, masking the problem.

To solve this, I transposed the s-vectors in cutIntro and changed their type from
List[List[FOLTerm]]
to
Set[List[FOLTerm].

The semantics are now in line with the cutIntro-paper: one entry of the set corresponds to a single term, and the entries inside a list are the values for the variables alpha_0,...,alpha_n.
